README.md
# CRPClustering: An R Package for Bayesian Nonparametric Chinese Restaurant Process Clustering with Entropy 

[![Build Status](https://travis-ci.org/jirotubuyaki/CRPClustering.svg?branch=master)](https://travis-ci.org/jirotubuyaki/CRPClustering)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/CRPClustering)](http://cran.r-project.org/package=CRPClustering)
[![](http://cranlogs.r-pkg.org/badges/grand-total/CRPClustering)](http://cran.rstudio.com/web/packages/CRPClustering/index.html)
## Now Under Construction
 Please wait a new version 1.4

## News 
1.4 (2020-04-12)  
  * Implemented by Rcpp. We changed main methods.  
  * "mu_0" and "sigma_table" and "ro_0" arguments were deleted in "crp_train" method.  
  * "data" argument of a matrix type was changed into data.frame type in "crp_train" method.  
  * return value in "crp_train" method was changed.  
  * A logical type "plot" argument in order to plot clustering result was added in "crp_train" method.   
  * "crp_plot_z" method was added.  
  
1.3 (2018-08-01)  
  * Implemented by Scala. We changed main methods.  
  

## Abstract
Clustering is a scientific method which finds the clusters of data and many related methods are traditionally researched. Bayesian nonparametrics is statistics which can treat models having infinite parameters. Chinese restaurant process is used in order to compose Dirichlet process. The clustering which uses Chinese restaurant process does not need to decide the number of clusters in advance. This algorithm automatically adjusts it. Then, this package can calculate clusters in addition to entropy as the ambiguity of clusters.

## Introduction
Clustering is an analytical method in order to find the clusters of data and many related methods are proposed. K-means[1] and Hierarchical clustering[2] are famous algorithmic methods. Density-based clustering[3] is the method that finds clusters by calculating a concentration of data. In statistical methods, there are stochastic ways such as bayesian clustering[4]. However these methods need to decide the number of clusters in advance. Therefore if the data is both high dimensions and a complex, deciding the accurate number of clusters is difficult. Bayesian nonparametric method[5] composes infinite parameters by Dirichlet process[6]. Dirichlet process is the infinite dimensional discrete distribution that is composed by Stocastic processes like a chinese restaurant process (CRP)[7] or stick-breaking process[8]. CRP does not need to decide the number of clusters in advance. This algorithm automatically adjusts it. We implement the CRP Clustering and the method which calculates the entropy[9] into R package. Then, we explain the clustering model and how to use it in detail and execute simulation by example datasets.

## Background
### Chinese Restaurant Process
Chinese restaurant process is a metaphor looks like customers sit at a table in Chinese restaurant. All customers except for x_i have already sat at finite tables. A new customer x_i will sit at either a table which other customers have already sat at or a new table. A new customer tends to sit at a table which has the number of customers more than other tables. A probability equation is given by    

![equa](./readme_images/equation_1.png "eque")

where n^i_k denotes the number of the customers at a table k except for i and α is a concentration parameter.

### Markov Chain Monte Carlo Methods for Clustering
Markov chain Monte Carlo (MCMC) methods[10] are algorithmic methods to sample from posterior distributions. If conditional posterior distributions are given by models, it is the best way in order to acquire parameters from posterior distributions. The algorithm for this package is given by  

i) ii) iterations continue on below:  

i) Sampling z_i for each i (i = 1,2, ・・・,n)

![equa](./readme_images/equation_2.png "eque")

where k is a k th cluster and i is a i th data. μ new and Σ new are calculated from all data.

ii) Calculating parameters for each k (k = 1,2, ・・・,∞)

![equa](./readme_images/equation_3.png "eque")

Iterations i) ii) continue by iteration number, and Σ k is a variance-covariance matrix of kth cluster. Cov is covariance. i and j are rows and columns’ number of Σ k ij and Σ new ij . First several durations of iterations which are called as burn_in are error ranges. For that reason, burn_in durations are abandoned.

### Clusters Entropy
Entropy denotes the ambiguity of clustering. As a result of a simulation, data x i joins in a particular cluster. From the total numbers n k of the particular cluster k at the last iteration, a probability p k at each cluster k is calculated. The entropy equation is given by

![equa](./readme_images/equation_4.png "eque")


## Installation
CRPClustering is available through GitHub (https://github.com/jirotubuyaki/CRPClustering) or CRAN (https://CRAN.R-project.org/package=CRPClustering). If download from GitHub, you can use devtools by the commands:

```
> library(devtools)
> install_github("jirotubuyaki/CRPClustering")
```

Once the packages are installed, it needs to be made accessible to the current R session by the commands:

```
> library(CRPClustering)
```

For online help facilities or the details of a particular command (such as the function crp_train) you can type:

```
> help(package="CRPClustering")
```

## Methods

### Implemented by C++ using GNU Scientific Library

This package is implemented by C++. Please install GNU Scientific Library. Programs are compiled. If you are interested in source codes by R and C++. Please read src directory.

### Method for Chinese Restaurant Process Clustering  

```
> result <- crp_train (data ,
                       alpha =1 ,
                       burn_in =100 ,
                       iteration =1000
                       plot = TRUE
                      )
```

This method calculates CRP clustering.  
Let arguments be:  
  ・ data : a data.frame of data for clustering. row is each data i and column is dimensions of each data i.  
  ・ alpha : a numeric of a CRP concentration rate.  
  ・ burn_in : an iteration integer of burn in.  
  ・ iteration : an iteration integer.  
  ・ plot : a logical type of whether plot a result or not.  
Let return be:  
  ・ result : a list has three elements. The "clusters" is cluster number and joined data number and cluster’s mean and variance matrix. The "max" is the cluster number for data i join in. The "z" is the iteration history for an each data i join in clusters.  
  
### Predict Which Cluster Data Join In  
This method predicts which cluster data join in.   

```
> predict <- crp_predict(data, result)
```
  
Let arguments be:  
  • data: a data.frame of data for clustering. row is each data i and column is dimensions of each data i.   
  • result: return from method "crp_train".  
Let return be:  
  • predict: an array denotes first colum is joined cluster and nexts are joined probability for each result cluster.

### Matrix Visualization For Predicting

```
> crp_plot(data, predict)
```

This method exhibits multi dimentional plot matrix. Let arguments be:  
  • data: a data.frame of data for clustering. Row is each data i and column is dimensions of each data i.  
  • predict: return from method "crp_predict".    

### Visualization of Clusters’ Probability Data i Join in.  

```
> crp_plot_z (i , result = result )
```

This method exhibits multi dimentional plot matrix. Let arguments be:  
  • i: a number of an each data i.  
  • result: return from crp_train method.    


## Simulation
We use dataset from Clustering basic benchmark( http://cs.joensuu.fi/sipu/datasets/ )[11]. If increase α parameter, new clusters tend to increase. burin_in iterations are abandoned. The result is plotted and each data joins in any cluster. The graph is given by below:

![equa](./readme_images/Rplot01.png "eque")

Figure 1: Aggregation: Data is 788 elements and 2 dimentions. Parameters are set as alpha=1, burnin=100, iteration=1000.  

![equa](./readme_images/Rplot02.png "eque")  

Figure 2: 3 normal distribution: Data is 1000 elements and 2 dimentions. Parameters are set as alpha=0.5, burnin=100, iteration=1000.  

![equa](./readme_images/Rplot.png "eque")  

Figure 3: 10 dimentional normal distributions: Data is generated from 10 dimentional normal distributions and parameters are set as alpha=1, burnin=100, iteration=1000.  

## Conclusions
Chinese restaurant process clustering was implemented and explained how to use it. Computer resources are limited. Computer processing power is the most important problem. After this, several improvements are planed. Please send suggestions and report bugs to okadaalgorithm@gmail.com.

## Acknowledgments
This activity would not have been possible without the support of my family and friends. To my family, thank you for much encouragement for me and inspiring me to follow my dreams. I am especially grateful to my parents, who supported me all aspects.  

## References
[1] Hartigan, J. A.; Wong, M. A. Algorithmas136: A k-means clustering algorithm . Journal of the Royal Statistical Society, Series C. 28 (1): 100–108. JSTOR 2346830, 1979.  

[2] Rokach, Lior, and Oded Maimon. "Clustering methods." Data mining and knowledge discovery handbook. Springer US, 2005. 321-352.  

[3] Ester, Martin; Kriegel, Hans-Peter; Sander, Jörg; Xu, Xiaowei (1996). Simoudis, Evangelos; Han, Jiawei; Fayyad, Usama M. (eds.). A density-based algorithm for discovering clusters in large spatial databases with noise. Proceedings of the Second International Con ference on Knowledge Discovery and Data Mining (KDD-96). AAAI Press. pp. 226–231.  

[4] John W Lau & Peter J Green (2007) Bayesian Model-Based Clustering Procedures, Journal of Computational and Graphical Statistics, 16:3, 526-558, DOI: 10.1198/106186007X238855.  

[5] Muller Peter, et al. Bayesian Nonparametric Data Analysis. Springer, 2015.  

[6] Ferguson, Thomas. Bayesian analysis of some nonparametric problems. Annals of Statistics. 1 (2): 209–230., 1973.  

[7] Pitman, Jim. Exchangeable and partially exchangeable random partitions. Probability Theory and Related Fields 102 (2): 145–158., 1995.  

[8] Broderick, Tamara, et al. “Beta Processes, Stick-Breaking and Power Laws.” Bayesian Analysis, vol. 7, no. 2, 2012, pp. 439–476., doi:10.1214/12-ba715.  

[9] Elliott H. Lieb; Jakob Yngvason. The physics and mathematics of the second law of thermodynamics. Physics Reports Volume:310 Issue:1 1-96., 1999.  

[10] Liu, Jun S. The collapsed gibbs sampler in bayesian computations with applications to a gene regulation problem. Journal of the American Statistical Association 89 (427): 958–966., 1994.  

[11] P. Fränti and S. Sieranoja K-means properties on six clustering benchmark datasets. Applied Intelligence, 48 (12), 4743-4759, December 2018.


NEWS
1.4 (2020-04-12)
  * Implemented by Rcpp. We changed main methods.

  * "mu_0" and "sigma_table" and "ro_0" arguments were deleted in "crp_train" method.

  * "data" argument of a matrix type was changed into data.frame type in "crp_train" 
  method.
  * return value in "crp_train" method was changed.

  * A logical type "plot" argument in order to plot clustering result was added in "crp_train" method.
  
  * "crp_plot_z" method was added.

1.3 (2018-08-01)
  * Implemented by Scala. We change main methods.


src/RcppExports.cpp
// Generated by using Rcpp::compileAttributes() -> do not edit by hand
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#include <RcppGSL.h>
#include <Rcpp.h>

using namespace Rcpp;

// gibbs
List gibbs(const int data_length, const int dim, const RcppGSL::matrix<double>& data_raw, const RcppGSL::matrix<double>& data_raw_result, const double alpha, const int burn_in, const int iteration);
RcppExport SEXP _CRPClustering_gibbs(SEXP data_lengthSEXP, SEXP dimSEXP, SEXP data_rawSEXP, SEXP data_raw_resultSEXP, SEXP alphaSEXP, SEXP burn_inSEXP, SEXP iterationSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< const int >::type data_length(data_lengthSEXP);
    Rcpp::traits::input_parameter< const int >::type dim(dimSEXP);
    Rcpp::traits::input_parameter< const RcppGSL::matrix<double>& >::type data_raw(data_rawSEXP);
    Rcpp::traits::input_parameter< const RcppGSL::matrix<double>& >::type data_raw_result(data_raw_resultSEXP);
    Rcpp::traits::input_parameter< const double >::type alpha(alphaSEXP);
    Rcpp::traits::input_parameter< const int >::type burn_in(burn_inSEXP);
    Rcpp::traits::input_parameter< const int >::type iteration(iterationSEXP);
    rcpp_result_gen = Rcpp::wrap(gibbs(data_length, dim, data_raw, data_raw_result, alpha, burn_in, iteration));
    return rcpp_result_gen;
END_RCPP
}

static const R_CallMethodDef CallEntries[] = {
    {"_CRPClustering_gibbs", (DL_FUNC) &_CRPClustering_gibbs, 7},
    {NULL, NULL, 0}
};

RcppExport void R_init_CRPClustering(DllInfo *dll) {
    R_registerRoutines(dll, NULL, CallEntries, NULL, NULL);
    R_useDynamicSymbols(dll, FALSE);
}


R/onLoad.R
#-----------------------------------------------------------------------Welcome
#               Chinese Restaurant Process Clustering
#                                                                    2020/04/12
#                                                                 Masashi Okada
#                                                      okadaalgorithm@gmail.com
#
#
#
# Description: Chinese Restaurant Process is used in order to compose Dirichlet
# Process. The Clustering which uses Chinese Restaurant Process does not need to
# decide a number of clusters in advance. This algorithm automatically adjusts
# it. And this package calculates ambiguity as entropy.

#---------------------------------------------------------------------------End

#-----------------------------------------------------------------------Library
library(Rcpp)
library(RcppGSL)
library(ggplot2)
library(GGally)
library(MASS)
library(matrixcalc)
#---------------------------------------------------------------------------End

#' Markov chain Monte Carlo methods for CRP Clustering
#' @import grid
#' @import png
#' @import Rcpp
#' @import RcppGSL
#' @import ggplot2
#' @import GGally
#' @importFrom utils read.table
#' @param data: a data.frame of data for clustering. Row is each data_i and column is dimensions of each data_i.
#' @param alpha : a numeric of a CRP concentrate rate.
#' @param burn_in :  an abandoned iteration integer.
#' @param iteration : an iteration integer.
#' @param plot : a logical type of whether plot a result or not.
#' @return result : a list has three elements. The "clusters" is cluster number and joined data number and cluster's mean and variance matrix. The "max" is the cluster number for data i join in. The "z" is the iteration history for each data i join in paticular cluster.
#' @examples
#' data <- array(0, dim=c(30, 2))
#' data[1, ] <- c(-2.43185475495409,2.03311203531621)
#' data[2, ] <- c(-0.408783769317068,1.75003135626229)
#' data[3, ] <- c(1.55675133967144,-1.56420523659905)
#' data[4, ] <- c(0.657368060264562,-0.97383866031164)
#' data[5, ] <- c(0.068212889245427,1.13418709295999)
#' data[6, ] <- c(-1.73561815639666,1.81918235905252)
#' data[7, ] <- c(1.03672457449158,-1.8569658734557)
#' data[8, ] <- c(1.76100672727452,0.0761038984873157)
#' data[9, ] <- c(1.65809552931208,-1.68283298969143)
#' data[10, ] < c(-1.07764013453211,-0.32226716632212)
#' data[11, ] <- c(-0.261606415434224,1.75394513146391)
#' data[12, ] <- c(-0.70752394485502,-0.259888201772335)
#' data[13, ] <- c(-1.38617236009739,-1.24305620615396)
#' data[14, ] <- c(1.51663782772836,0.0161396983547837)
#' data[15, ] <- c(-0.914042151747574,1.76495756094862)
#' data[16, ] <- c(0.282796296711756,-0.0492279088948679)
#' data[17, ] <- c(1.08831769386705,-0.954851525684512)
#' data[18, ] <- c(-0.932745904717591,0.762387679372797)
#' data[19, ] <- c(1.69617665862324,-1.11969182200371)
#' data[20, ] <- c(-0.767903781929682,1.19342570049071)
#' data[21, ] <- c(0.401780436172324,-0.457100405625718)
#' data[22, ] <- c(0.98090840279728,-0.597487493647301)
#' data[23, ] <- c(-1.29713416781756,0.765401326146141)
#' data[24, ] <- c(-1.7620625656782,2.02686171889867)
#' data[25, ] <- c(1.68478051448753,-0.806918914294913)
#' data[26, ] <- c(0.466622923887724,0.197650126048092)
#' data[27, ] <- c(1.4851646799543,-1.53289806708663)
#' data[28, ] <- c(-2.12370907063395,1.6471140089684)
#' data[29, ] <- c(-0.660332309091402,1.73989289688085)
#' data[30, ] <- c(-0.957127602683051,-0.0156523076019355)
#' data <- round(data, 3)
#' result <- crp_train(
#'                       data,
#'                       alpha=1,
#'                       burn_in=10,
#'                       iteration=100,
#'                       plot=TRUE
#'                      )
#' @export
crp_train<- function(data = data, alpha=1, burn_in=100, iteration=1000, plot=TRUE) {
  if(is.numeric(alpha) == FALSE){
    print("Error: Input alpha type is not numeric.")
    return(FALSE)
  }
  if(alpha <= 0){
    print("Error: Input alpha is less than or equal to zero.")
    return(FALSE)
  }
  if(burn_in <= 0){
    print("Error: Input burn_in is less than or equal to zero.")
    return(FALSE)
  }
  if(iteration <= 0){
    print("Error: Input iteration is less than or equal to zero.")
    return(FALSE)
  }
  if(iteration <= burn_in){
    print("Error: Input iteration is more than burn_in.")
    return(FALSE)
  }
  if(is.logical(plot) == FALSE){
    print("Error: Input plot type is not logical.")
    return(FALSE)
  }
  data_result <- as.matrix(data)
  data <- scale(data)
  data <-as.matrix(data)

  data_length <- nrow(data)
  dim <- ncol(data)
  burn_in <- as.integer(burn_in)
  iteration <- as.integer(iteration)
  out <- gibbs(data_length, dim, data,data_result, alpha, burn_in, iteration)
  clusters <- out$clusters
  joined <- out$result
  z <- as.matrix(out$z)
  z_output <- z[burn_in:iteration, 1:data_length]
  result_frame <- data.frame(ClusterNumber = clusters[ , 1], TotalNumber = clusters[ , 2] ,ResultMean = clusters[ , 3 : (3 + dim - 1)], ResultVariance = clusters[ , (3 + dim) : (3 + dim * 2 + 1)])
  output <- list(result = result_frame, z = z_output)

  print(result_frame)
  
  if(plot == TRUE){
    dim <- ncol(data_result)
    data_result_frame <- data.frame(data_result,cluster=joined)
    p_ <- GGally::print_if_interactive
    map <- ggpairs(data_result_frame, columns = c(1:dim),
                   diag=list(mapping = aes(color = factor(cluster),alpha=0.125), continuous=wrap("barDiag", binwidth=0.3)),
                   lower = list(mapping = aes(color = factor(cluster)), continuous = wrap("points", size=0.42,stroke = 0.3)),
                   upper = list(mapping = aes(color = factor(cluster), alpha=0.6), continuous = wrap("density", size=.35))
    )
    p_(map)
    remove(p_)
    remove(map)
    remove(data_result_frame)
  }
  remove(alpha)
  remove(burn_in)
  remove(iteration)
  remove(data_length)
  remove(dim)
  remove(plot)
  remove(out)
  remove(clusters)
  remove(joined)
  remove(z)
  remove(z_output)
  remove(result_frame)
  remove(data)
  remove(data_result)

  return(output)
}
#' Predict The Clusters' Probability that Data Joins in. 
#' @import mvtnorm
#' @import MASS
#' @import matrixcalc
#' @param data: a data.frame of data for clustering. Row is each data_i and column is dimensions of each data_i.
#' @param result: a list returned from crp_train method.
#' @return predict: an array expresses a joined cluster number and joined probability in each cluster. 
#' @export
crp_predict<- function(data = data, result = result){
  data <- as.matrix(data)
  result_result <- result$result
  result_result <- as.matrix(result_result)
  dim <- ncol(data)
  data_length <- nrow(data)
  cluster_length <- nrow(result_result)
  


  predict_result <- array(0,dim=c(data_length, cluster_length + 1))
  cluster_max <- array(0,dim=c(data_length))
  cluster_max_value <- array(0,dim=c(data_length))
  name <- seq(0, length= cluster_length + 1)
  name[1] = "ClusterNumber"
  for(i in 1 : cluster_length){
     name[i + 1] = paste("ClusterProbability-", result_result[i, 1]) 
  }
  colnames(predict_result) <- name 

  count = 1
  for(i in 1 : data_length){
    for(k in 1 : cluster_length){
      mu <- result_result[k, 3 : (3 + dim -1)]
      sigma <- matrix(result_result[k, (3 + dim) : (2 + dim + dim * dim )], nrow = dim, ncol = dim)
      predict_result[i, count + 1] <-  dmvnorm(data[i,], mu, sigma)
      if(cluster_max_value[i] <= predict_result[i, count + 1]){
        predict_result[i, 1] <- result_result[count, 1]
        cluster_max_value[i] <- predict_result[i, count + 1]
      }
      count = count + 1  
    }
    count <- 1
  }
  remove(cluster_max)
  remove(cluster_max_value)
  remove(mu)
  remove(sigma)
  remove(count)
  remove(data)
  remove(result)
  remove(result_result)
  remove(cluster_length)
  remove(data_length)
  remove(dim)
  remove(i)
  remove(k)

  return (predict_result)
}

#' CRP Plot Matrix Visualization
#' @import ggplot2
#' @import GGally
#' @param data: a data.frame of data for clustering. Row is each data_i and column is dimensions of each data_i.
#' @param predict : an array returned from crp_predict method.
#' @export
crp_plot <- function(data = data, predict =  predict){
  data <- as.matrix(data)
  dim <- ncol(data)
  data_frame <- data.frame(data,cluster=predict[ , 1])
  p_plot_ <- GGally::print_if_interactive
  map_plot <- ggpairs(data_frame, columns = c(1:dim),
                 diag=list(mapping = aes(color = factor(cluster), alpha=0.125), continuous=wrap("barDiag", binwidth=0.3)),
                 lower = list(mapping = aes(color = factor(cluster)), continuous = wrap("points", size=0.42,stroke = 0.3)),
                 upper = list(mapping = aes(color = factor(cluster), alpha=0.6), continuous = wrap("density", size=.35))
  )
  p_plot_(map_plot)
  remove(dim)
  remove(data)
  remove(data_frame)
  remove(predict)
  remove(p_plot_)
  remove(map_plot)
}
#' Visualization Probability of the Clusters that Data i Join in.
#' @import ggplot2
#' @param i: a number of each data i.
#' @param result: a list returned from crp_train method. 
#' @export
crp_plot_z <- function(i = 1, result =  result){
  z <- result[["z"]]
  z_i <- matrix(0, nrow=max(z[, i]) + 1 , ncol=1)
  for(j in 1 : nrow(z)){
    z_i[z[j, i] + 1, 1] = z_i[z[j, i] + 1, 1] + 1;
  }
  count <- 1
  for(j in 1 : nrow(z)){
    if(z_i[z[j, i] + 1, 1] != 0){
      count = count + 1
    }
  }
  z_label <- matrix(0, nrow=count, ncol=1)
  z_result <- matrix(0, nrow=count, ncol=1)
  count <- 1
  for(j in 1 : nrow(z_i)){
    if(z_i[j, 1] != 0){
      z_label[count, 1] <- j - 1 
      z_result[count, 1] = z_i[j, 1]
      count <- count + 1
    }
  }
  df <- data.frame(cluster = z_label[1:count, 1], count = z_result[1:count, 1])
  g <- ggplot(df, aes(x = factor(cluster), y = count)) 
  g <- g + geom_bar(stat = "identity", aes(colour=factor(cluster), fill=factor(cluster)))
  plot(g)
  remove(count)
  remove(z_label)
  remove(z_result)
  remove(df)
  remove(g)
  remove(result)
}










